backends/cxxrtl/runtime/cxxrtl/cxxrtl.h

```diff
@@ -871,8 +894,14 @@ struct wire {
 		next.template set<IntegerT>(other);
 	}
 
-	bool commit() {
+	// This method intentionally takes a mandatory argument (to make it more difficult to misuse in
+	// black box implementations, leading to missed observer events). It is generic over its argument
```

Comment:
> (to make it more difficult to misuse in black box implementations, leading to missed observer events)

I like this choice.

Reply:
I will most likely completely remove user-defined black boxes as they are right now, but in the interim this works.

Reply:
FWIW I have used user-defined black boxes and would be sad to see them go

Reply:
It's very difficult, bordering on impossible, to combine user-defined black boxes and record/replay debugging. They are not being completely removed though but rather replaced with something that is behaving more predictably.

The replacement will restrict the black boxes to only have synthesizable logic and also wires that are treated as-if they were toplevel ports, and provide a way to retrieve these wires in a way that is more convenient than individually plucking them from the debug items map.

The idea is that you put as much logic as you can into the synthesizable part (e.g. an UART receiver), and then, from the outside (from the simulation point of view, not necessarily C++ code point of view), you will be able to check a flag indicating there's a character received, and print it on screen. In most cases, these pseudo toplevel ports should be structured as FIFOs but there is freedom for the application to do whatever it wants, allowing implementing things like memories or caches as well.

This also avoids the race condition issue that has been plaguing the black box implementation since it was added.

The actual interface for instantiating these new-style black boxes isn't finalized yet.

<hr>

The above covers the use of black boxes for I/O. There is another valid use of black boxes, which is to make a simulation faster by replacing a block with the equivalent native code. I might continue to enable that use by removing the black box instantiation mechanism and simply emitting a normal module where `eval` is user-defined. `commit` would not be overridable at all, nor would instantiation be customizable.